### ISSUE TITLE
[Feature Request] Where clause alternative: "Composite Traits" `TraitVariant` (`|`) and `TraitUnion` (`&`)

### ISSUE THREAD

**Issue by martinvuyk** [modularml/mojo#3252]
### Review Mojo's priorities

- [X] I have read the [roadmap and priorities](https://docs.modular.com/mojo/roadmap.html#overall-priorities) and I believe this request falls within the priorities.

### What is your request?

Having the flexibility to define composite traits as aliases or directly in function signatures, adding to the meta-programmability of the language instead of using Rust style where clauses.

### What is your motivation for this change?

say you have 2 cases:
```mojo
trait HasRepr(Representable, CollectionElement):
    ...
trait HasStr(Stringable, CollectionElement):
    ...

struct List:
    ...
    fn __str__[U: HasRepr](self: List[U]) -> String:
        ...
    fn __str__[U: HasStr](self: List[U]) -> String:
        ...
```
what happens when a type implements both?

using `TraitVariant`:
```mojo
trait HasRepr(Representable, CollectionElement):
    ...
trait HasStr(Stringable, CollectionElement):
    ...

alias CanShow = HasRepr | HasStr

struct List:
    ...
    fn __str__[U: CanShow](self: List[U]) -> String:
        if U.implements[HasRepr]():
            ...
        else:
            ...
```

and using `TraitUnion`:
```mojo
struct List:
    ...
    fn __str__[
        U: CollectionElement & (Representable | Stringable)
    ](self: List[U]) -> String:
        if U.implements[Representable]():
            ...
        else:
            ...
```

### Any other details?

This could fix issues like #3148 by having print accept types that implement any one of: `Formattable`, `Representable`, `Stringable`
```mojo
fn print[
    *Ts: Formattable | Representable | Stringable
](
    *values: *Ts,
    sep: StaticString = " ",
    end: StaticString = "\n",
    flush: Bool = False,
    file: FileDescriptor = stdout,
):
    # the implementation uses Formattable if available since it is more efficient
    # falling back to Representable if the type implements it
    # otherwise using Stringable trait
    ...
```

**Comment by soraros** [comment]
This method of proceeding suffers from the same defect as `rebind`: namely, the type is not automatically refined within `@parameter if` branches.

```mojo
struct List:
  ...
  fn __str__[U: CanShow](self: List[U]) -> String:
    var a = self[0]
    @parameter
    if U.implements[HasRepr]():
      # the compiler will not resolve a.__repr__()
    else:
      ...
```

One might reasonably contend that the compiler ought to refine types under such conditions. However, this leads immediately to a deeper difficulty: there always exist functions of type `fn[...]() -> Bool` whose outcomes the compiler cannot symbolically simplify. For example:

```mojo
# context:
alias n: Int
alias is_prime: fn[n: Int]() -> Bool
alias has_str: fn[T]() -> Bool

@parameter
if U.implements[HasRepr]() and is_prime[n]():
  # the compiler will not resolve a.__repr__()
elif U.implements[HasRepr]() and not is_prime[n]():
  # we cannot eliminate `is_prime`; do we leave it unevaluated?
  ...
elif U.implements[HasRepr]() and has_str[U]():
  # If we answered "yes" to the previous question, what do we do here?
  # How would the compiler know, without unfolding `has_str`, that this case permits refinement?
  ...
else:
  # Failure to resolve any of the above cases leaves the `else` branch ill-formed,
  # since type refinement operates via `U - (br1 ∨ br2 ∨ ...)`
  ...
```

At bottom, there are two fundamentally distinct kinds of predicates involved.  
First, there are predicates whose truth can be (or can be made to) mechanically determined in a generic context, such as type equality, trait conformance, or structural properties of types (`isinstance`, your `implements`, `_type_is_eq`, etc.). These can drive type refinement, because the compiler has sufficient local information to decide their validity without executing arbitrary code.  
Second, there are predicates whose truth cannot, in general, be determined symbolically, such as runtime properties, opaque function calls, or semantically undecidable conditions (`is_prime`, `size.is_power_of_two()` in `SIMD`, etc.). These cannot meaningfully trigger refinement, as doing so would require solving arbitrary symbolic computation problems at compile time.

The `@parameter if` construct, however, permits an arbitrary mix of these two classes of predicates. Once mixed, the compiler has no systematic way to distinguish "refinable" from "non-refinable" cases without a full semantic analysis of user code — an undecidable and practically infeasible task. Hence, the problem is, in essence, unsolvable.

**Comment by martinvuyk** [comment]
@soraros I agree that the compiler might not need to resolve those cases (if they're even solvable) because it would also bloat compile times.

```mojo
@parameter
if U.implements[HasRepr]() and is_prime[n](): ...
```
I imagine there can be some workarounds like e.g. not letting implements be a compile time branch, or not letting them exist in the same branch as a computed value, or forcing them to be standalone in each branch (making them an mlir bool value that doesn't implement an and/or), etc.

As for this:
```mojo
if U.implements[HasRepr]() and is_prime[n]():
  # the compiler will not resolve a.__repr__()
```
It's why I also asked for a `rebind_trait` in #3630
